LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/UdpServer.cs

```diff
@@ -22,12 +24,12 @@ public class UdpServer : IDisposable
         private readonly NetworkServerConfiguration configuration;
         private readonly LoraDeviceInfoManager loraDeviceInfoManager;
         ModuleClient ioTHubModuleClient;
+        private IPAddress remoteLoRaAggregatorIp;
+        private int pullAckRemoteLoRaAggregatorPort=0;
```

Comment:
I understand what you mean with keeping the port numbers.
Push and pull have different ports and you need to use the correct one based on the payload we are sending.

Reply:
I need to keep at least the variable pullAckRemoteLoRaAggregatorPort as a variable, as it is transmitted by a PULL_ACK. If I want to send PULL_RESP (downstream message) after receiving a PUSH DATA (upstream message), I would need to send it on this port, that i can only have if I save it. 
I could save the variable pushDataRemoteLoRaAggregatorPort as I only need it to answer to Push_Data answer.

Reply:
At least in the code there is a chance that you call UdpSendMessage using pullAckRemoteLoRaAggregatorPort having value 0

Scenario:

1. Module starts
2. PUSH_DATA request arrives
    The Task.Run could call UdpSendMessage with pullAckRemoteLoRaAggregatorPort being zero.

Could this happen?

Reply:
I don't think so, the goal of the PULL_DATA is to check that network route is open. If there are no reply, the packet forwarder would think the connection is closed. Here are some quotes from the specs that makes me think like this: 
This data exchange is initialized by the gateway because it might be 
impossible for the server to send packets to the gateway if the gateway is 
behind a NAT.

When the gateway initialize the exchange, the network route towards the 
server will open and will allow for packets to flow both directions.
The gateway must periodically send PULL_DATA packets to be sure the network 
route stays open for the server to be used at any time.